esx_service/vmdk_ops.py

```diff
@@ -1786,11 +1824,16 @@ def handleVmciRequests(port):
     cartel = c_int32()
     vmci_grab_listening_socket(port)
 
-
     while True:
+        # Check the flag and the counter - exit the loop if flag is set and counter equals 0
+        if (stopEvent.isSet() and (opsCounter.value == 0)):
+            logging.info("Stopping service: no operations in flight.")
+            break;
+
+        # Listening on VMCI socket
```

Comment:
so if there is an operation in flight, you fall through to like 1835 and start waiting on vmci. Until vmci does not return,  the code is stuck and will not exit. Seems like a bug to me

Reply:
Agree. Looks like we should check stopEvent.isSet() only and remove the check (opsCounter.value == 0). But I guess we should wait until all in-flight ops done before invoking vmci_release_listening_socket() at line 1874? @msterin

Reply:
yes, release_listening_socket needs to wait untill all is done. 
Generally.  the way I expected the code was (in pseudo-code) something like this: 

```python
lock= threading.Rlock() # used for counters
condition = threading.Condition(lock) # used to signal on the counter drop
barrier = False

    signal_handler: 
        barrier = True
        spawn a waiting_thread
        return 

   main_thread:
        while get_vmci_op:
           if (barrier) vmci_reply("shutting down"); continue
           increase count
           execute_thread_for_command

thread_for_command:
    count.decr() on exit

Count:
   decr:
        with Rlock()  
          count -= 1
          if (count == 0) :
                 condition.notifyAll()


   incr:
       if (barrier): 
         assert_or_yell
      with Rlock:
       count += 1

waiting_thread":
      condition.acquire()
     try:  
        condition.wait(timeout)
    except RuntimeError:
         log.warning("dropped the wait due to timeout")
     sys.exit()
 ```

or you can use threading.event() to signal/wait instead of condition (events are simpler) 
In the current code, it's harder to navigate the logic for issues ....

Reply:
Thanks for sharing your thoughts. Will try to address in follow-up changes.